homeassistant/components/aprs/device_tracker.py

```diff
@@ -0,0 +1,186 @@
+"""Support for APRS device tracking."""
+
+import logging
+import threading
+
+import voluptuous as vol
+
+from homeassistant.components.device_tracker import (
+    ATTR_GPS_ACCURACY, ATTR_LATITUDE, ATTR_LONGITUDE, PLATFORM_SCHEMA)
+from homeassistant.const import (
+    CONF_HOST, CONF_PASSWORD, CONF_TIMEOUT, CONF_USERNAME,
+    EVENT_HOMEASSISTANT_STOP)
+from homeassistant.exceptions import PlatformNotReady
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util import slugify
+
+DOMAIN = 'aprs'
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_ALTITUDE = 'altitude'
+ATTR_COURSE = 'course'
+ATTR_COMMENT = 'comment'
+ATTR_FROM = 'from'
+ATTR_FORMAT = 'format'
+ATTR_POS_AMBIGUITY = 'posambiguity'
+ATTR_SPEED = 'speed'
+
+CONF_CALLSIGNS = 'callsigns'
+
+DEFAULT_HOST = 'rotate.aprs2.net'
+DEFAULT_PASSWORD = '-1'
+DEFAULT_TIMEOUT = 30.0
+
+FILTER_PORT = 14580
+
+MSG_FORMATS = ['compressed', 'uncompressed', 'mic-e']
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_CALLSIGNS): cv.ensure_list,
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Optional(CONF_PASSWORD,
+                 default=DEFAULT_PASSWORD): cv.string,
+    vol.Optional(CONF_HOST,
+                 default=DEFAULT_HOST): cv.string,
+    vol.Optional(CONF_TIMEOUT,
+                 default=DEFAULT_TIMEOUT): vol.Coerce(float)
+})
+
+
+def make_filter(callsigns: list) -> str:
+    """Make a server-side filter from a list of callsigns."""
+    return ' '.join('b/{0}'.format(cs.upper()) for cs in callsigns)
+
+
+def gps_accuracy(gps, posambiguity: int) -> int:
+    """Calculate the GPS accuracy based on APRS posambiguity."""
+    import geopy.distance
+
+    pos_a_map = {0: 0,
+                 1: 1 / 600,
+                 2: 1 / 60,
+                 3: 1 / 6,
+                 4: 1}
+    if posambiguity in pos_a_map:
+        degrees = pos_a_map[posambiguity]
+
+        gps2 = (gps[0], gps[1] + degrees)
+        dist_m = geopy.distance.distance(gps, gps2).m
+
+        accuracy = round(dist_m)
+    else:
+        message = "APRS position ambiguity must be 0-4, not '{0}'.".format(
+            posambiguity)
+        raise ValueError(message)
+
+    return accuracy
+
+
+def setup_scanner(hass, config, see, discovery_info=None):
+    """Set up the APRS tracker."""
+    callsigns = config.get(CONF_CALLSIGNS)
+    server_filter = make_filter(callsigns)
+
+    callsign = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    host = config.get(CONF_HOST)
+    timeout = config.get(CONF_TIMEOUT)
+    aprs_listener = AprsListenerThread(
+        callsign, password, host, server_filter, see)
+
+    def aprs_disconnect(event):
+        """Stop the APRS connection."""
+        aprs_listener.stop()
+
+    aprs_listener.start()
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, aprs_disconnect)
+
+    if not aprs_listener.start_event.wait(timeout):
+        raise PlatformNotReady("Timeout waiting for APRS to connect.")
```

Comment:
This won't work for device tracker, because it is using older infrastructure than the `async_setup_platform` code for which the `PlatformNotReady` exception is meant.

Reply:
I'm ready to clean it up but things may have shifted in the interim -- any suggestions are appreciated.

Reply:
Print an error and return. In a future PR we can migrate this to config entries and you can opt-in to the new way the device tracker does things, using entities. It's a bit more complicated and I don't want this PR to blow up but instead get it merged. Got side tracked but still reviewing